content/00.build/40.tooling/30.foundry/40.migration-guide/30.testing.md

```diff
@@ -0,0 +1,390 @@
+    }
+}
+```
+
+In this fixed example, `msg.sender` is used instead of `tx.origin`, which ensures compatibility with ZKsync.
+
+### Reserved Address Range
+
+#### Issue
+
+On ZKsync's zkEVM, addresses in the range `[0..2^16-1]` are reserved for kernel space,
+meaning they are not available for contract deployment or other user-level interactions.
+Attempting to use these addresses, either directly or through mocking, can lead to undefined behavior in your tests.
+
+#### Problematic Code
+
+In this test example, the address `0` is used in a mocked call. Since this address is
+part of the reserved range, using it will result in undefined behavior and test failures.
+
+```solidity
+contract FooTest is Test {
+    function testFoo() public {
+        // Invalid: Using an address within the reserved range
+        vm.mockCall(
+            address(0),
+            abi.encodeWithSelector(bytes4(keccak256("number()"))),
+            abi.encode(5)
+        );
+    }
+}
+```
+
+This example attempts to mock a call to `address(0)`, which is part of the reserved kernel space and therefore invalid on ZKsync.
+
+#### Error
+
+Using addresses within the reserved range may lead to undefined behavior and possible
+test failures without specific error messages. It's crucial to avoid these addresses entirely to ensure test reliability.
+
+#### Solution
+
+Ensure that addresses used for testing and mocking are outside of the reserved kernel
+space range. Start from address `65536` onwards for any user-level interaction.
+
+In fuzz testing, you can exclude the reserved address range either by setting assumptions or using specific fuzz configuration.
+
+Here’s a corrected test example:
+
+#### Fixed Code Example
+
+```solidity
+contract FooTest is Test {
+    function testFoo() public {
+        // Valid: Using an address outside of the reserved range
+        vm.mockCall(
+            address(65536),
+            abi.encodeWithSelector(bytes4(keccak256("number()"))),
+            abi.encode(5)
+        );
+    }
+}
+```
+
+In this corrected example, we use `address(65536)`, which is a valid user-space address in ZKsync.
+
+#### Additional Configuration
+
+During fuzz testing, it's important to avoid generating addresses within the reserved range. This can be done in two ways:
+
+1. **Using `vm.assume`**:
+   You can assert that the address generated in fuzz testing is greater than or equal to `65536`:
+
+   ```solidity
+   vm.assume(address(value) >= 65536);
+   ```
+
+2. **Fuzz Configuration**:
+   Alternatively, you can set the `no_zksync_reserved_addresses` option in your fuzz configuration to automatically exclude addresses in the reserved range.
+
+   ```toml
+   [profile.default]
+   fuzz = { no_zksync_reserved_addresses = true }
+   ```
+
+These approaches ensure that fuzz testing
+avoids the reserved kernel space,
+preventing test failures and ensuring compliance with ZKsync's address rules.
```

Comment:
@itsacoyote weird. I just pushed a commit ensuring there was a new line but its not showing up here? 

![Screenshot 2024-09-18 at 9 08 09 AM](https://github.com/user-attachments/assets/b94754d2-a338-4dcb-b762-a122e4c5a44b)

Reply:
It seems to pass the CI build and linting tho?

Reply:
🤷🏻 if it's not breaking any of our CI and linting, let's leave it then 👍🏻